src/context.rs

```diff
@@ -0,0 +1,51 @@
+use crate::{credentials::Credentials, package::PackageStore};
+use miette::{IntoDiagnostic, Result};
+use std::path::PathBuf;
+use std::sync::Arc;
+
+/// Shared context representing `buffrs` project.
+#[derive(Debug)]
+pub struct Context {
+    /// Package store, contains installed dependencies.
+    store: PackageStore,
+    /// Credentials, used to access registries.
+    credentials: Arc<Credentials>,
+}
+
+impl Context {
+    /// Package store
+    pub fn store(&self) -> &PackageStore {
```

Comment:
Can we make the fields public rather than using getters?

Reply:
Context: If we access this through an arc, we have the guarantee that no one mutates it anyway

Reply:
Yes we can. I find that using getters here is more clean however. This is not a *plain old data* struct that we expect to be stable and not change. We want to be able to change the layout here if we want to without touching a million files.

Reply:
I think this is not as bad as you make it sound, if we change the name of "store" to something else, we would likely also rename the getter for consistency. 

Bulk renaming is not a bad thing imo - and if we don't want to change the api, we would just not rename anything (is my guess)

Reply:
It is up to you, but here are some concerns:

- An is not a guarantee for read-only semantics: [`Arc::get_mut()`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.get_mut).
- This `Context` type is not just a loose bag of data, it has a semantic meaning. It represents one "validated" context of a `buffrs` project. Making the fields public means you can at any point mutate them, or in the case of an `Arc`, override them. Usually, when we build abstractions like this, we don't want to make fields public to prevent the type from being misused. All access should go through explicit member functions and it should only be constructed through explicit constructor methods. 
- As for refactoring: even if the refactoring itself is not so bad, imagine a project that has a healthy number of contributors and about 30 in-progress branches. Now, you need not only refactor your local stuff but you're going to have people yelling at you because all of their work is broken because they need to apply this change in each of those branches. Refactoring is hard to scale, so usually it is easier to have some member methods to serve as the API. There is a saying that summarizes this: *be conservative in the API, but be liberal in the implementation*. By using non-`pub` fields here, we give ourselves the maximum of freedom to change things in the future.

In my personal code, I always differentiate between:

- *plain old data* structs. These typically derive `Serialize` and `Deserialize`. They have all-public fields. The reason that is okay is that these represent external data such as configuration, or API responses. They are guaranteed not to change and the types themselves do not serve as abstractions or encode any information.
- everything else: Implements some kind of an abstraction. All interaction through methods. Never implement `Serialize` and `Deserialize`. 

In fact, this rule is so common that there is a `clippy` lint for it called [`partial_pub_fields`](https://rust-lang.github.io/rust-clippy/master/index.html#/partial_pub_fields), I recommend you read the description for it. I wish it did a better job of explaining the background reasoning of it.